ext_proc: convert ext_proc filter into dual filter

yanjunxiang-google opened the pull request:
This PR converts ext_proc filter into dual filter so it can be created in the upstream filter chain.

<!--
!!!ATTENTION!!!

If you are fixing *any* crash or *any* potential security issue, *do not*
open a pull request in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.
Thank you in advance for helping to keep Envoy secure.

!!!ATTENTION!!!

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)
-->

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

yanjunxiang-google commented:
/assign @tyxia

tyxia commented:
@yanjunxiang-google  Thanks working on this!

Before review starts, could you please fix the format issue so that the all presubmit tests can run?

yanjunxiang-google commented:
Updated the format

> @yanjunxiang-google Thanks for working on this!
> 
> Before review starts, could you please fix the format issue so that the all presubmit tests can run?

Updated the format. One thing is that with ext_proc filter moving to the upstream filter chain, as it is emitting gRPC call stats here: https://github.com/envoyproxy/envoy/blob/4148fb75513fbc34dbdad53fca538c7c75e3334e/source/extensions/filters/http/ext_proc/ext_proc.cc#L220.   The consumer of the gRPC call stats, like some logging filters,  probably needs to be created in the upstream filter chain after it as well.

yanjunxiang-google commented:
Another thing is that if side stream server ask Envoy to send immediate response, which will trigger ext_proc filter to send a localReply to downstream client, can ext_proc filter do that when it is in upstream filter chain?  Maybe we should take the immediate response as suspicious if the ext_proc filter is in upstream filter chain.

tyxia commented on source/extensions/filters/http/ext_proc/config.cc:
```diff
@@ -9,62 +9,41 @@ namespace Extensions {
   return std::make_shared<FilterConfigPerRoute>(proto_config);
 }
 
-Http::FilterFactoryCb
```
Looks like the PR is still in draft mode. I just want to leave an early comment about removing this method.

Currently, it is used in match_delegate filter [like here](https://github.com/envoyproxy/envoy/blob/7c410102b1a97b439c8b86bc10c0c5059a408ca3/source/extensions/filters/http/composite/action.cc#L75) so that ext_proc filter can be created in match_delegate's per route case.

Thus, deleting this method will break that. We will need to come up with non-breaking solution along with recent refactor of factory/server context.

yanjunxiang-google commented:
/wait

yanjunxiang-google commented:
/re-open
